### PR TITLE
Add Android commandline tools

### PR DESCRIPTION
**Pull request by tir38**
**Important:** *Do not tick a checkbox if you haven’t performed its action.* Honesty is indispensable for a smooth review process.

After making all changes to a cask, verify:

- [x] The submission is for [a stable version](https://docs.brew.sh/Acceptable-Casks#stable-versions) or [documented exception](https://docs.brew.sh/Acceptable-Casks#but-there-is-no-stable-version).
- [x] `brew audit --cask {{cask_file}}` is error-free.
- [x] `brew style --fix {{cask_file}}` reports no offenses.

Additionally, **if adding a new cask**:

- [x] Named the cask according to the [token reference](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/token_reference.md).
- [x] Checked the cask was not [already refused](https://github.com/Homebrew/homebrew-cask/search?q=is%3Aclosed&type=Issues).
- [x] Checked the cask is submitted to [the correct repo](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#finding-a-home-for-your-cask).
- [x] `brew audit --new-cask {{cask_file}}` worked successfully.
- [x] `brew install --cask {{cask_file}}` worked successfully.
- [x] `brew uninstall --cask {{cask_file}}` worked successfully.

### CODE REVIEW

**Comment by miccal** [comment]
Previous attempt: https://github.com/Homebrew/homebrew-cask/pull/80305

**Comment by tir38** [comment]
Running `audit`, I don't understand `at least one activatable artifact stanza is required`

**Comment by SMillerDev** [comment]
> Running audit, I don't understand at least one activatable artifact stanza is required

you're currently not installing anything. You're just downloading a zip.

https://docs.brew.sh/Cask-Cookbook#at-least-one-artifact-stanza-is-also-required

File: Casks/android-commandlinetools.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+cask "android-commandlinetools" do
+  version "6858069"
+  sha256 "58a55d9c5bcacd7c42170d2cf2c9ae2889c6797a6128307aaf69100636f54a13"
+
+  url "https://dl.google.com/android/repository/commandlinetools-mac-#{version}_latest.zip",
+      verified: "dl.google.com"
+
```
Comment (cho-m) [inline Casks/android-commandlinetools.rb:7]: Please reorder items based on https://docs.brew.sh/Cask-Cookbook#stanza-order

File: Casks/android-commandlinetools.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  livecheck do
+    url "https://developer.android.com/studio#cmdline-tools"
```
Comment (cho-m) [inline Casks/android-commandlinetools.rb:9]:
```suggestion
    url "https://developer.android.com/studio#cmdline-tools"
    strategy :page_match
    regex(%r{href=.*?/commandlinetools-mac-(\d+)_latest\.zip}i)
```

File: Casks/android-commandlinetools.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+  name "android-cmdline-tools"
+  homepage "https://developer.android.com/studio"
+  desc "Android command-line tools without needing Android Studio"
+
+  conflicts_with cask: [
+    "android-sdk",
+    "android-platform-tools",
+  ]
+end
```
Comment (cho-m) [inline Casks/android-commandlinetools.rb:20]: You can reference previous https://github.com/Homebrew/homebrew-cask/pull/80305/files for using `binary` stanza.

The path may need to be changed if ZIP structure changed.

**Comment by cho-m** [comment]
I fixed up Cask to pass CI checks, but this shouldn't be merged until a user tests the installed tools are actually working.

Based on previous https://github.com/Homebrew/homebrew-cask/pull/80305#issuecomment-611771909, there were some issues with location of installation, but it may have been fixed in later version.

**Comment by tir38** [comment]
Thanks for the updates. I'll test this week.

**Comment by tir38** [comment]
@cho-m I confirmed that your changes work (thanks). Everything is passing. The binary mappings you add work fine. Here's where things get complicated

Brew installs sdkmanager (and others) in in Caskroom:

```
$ brew info android-commandlinetools
android-commandlinetools: 7302050
https://developer.android.com/studio
/opt/homebrew/Caskroom/android-commandlinetools/7302050 (95 files, 99.1MB)
==> Name
android-cmdline-tools
==> Description
Command-line tools for building and debugging Android apps
==> Artifacts
/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/bin/apkanalyzer (Binary)
/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/bin/avdmanager (Binary)
/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/bin/lint (Binary)
/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/bin/retrace (Binary)
/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/bin/screenshot2 (Binary)
/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/bin/sdkmanager (Binary)
```

The problem is that most apps (including Android Studio and sdkmanager itself) are expecting these to be in `/usr/USERNAME/Library/Android/sdk`


```
$ sdkmanager --list
Error: Could not determine SDK root.
Error: Either specify it explicitly with --sdk_root= or move this package into its expected location: <sdk>/cmdline-tools/latest/
```


I can specify `--sdk_root` every time:

```
$ $ sdkmanager --list --sdk_root=/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools
...
$ sdkmanager "sources;android-30" --sdk_root=/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools
...
$ /opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/sources/android-30
```

I don't think this scales for most people.  I think Android Studio will not be happy with this unless we set ANDROID_HOME (and ANDROID_SDK) env vars to this Caskroom location. Which I also don't think is a good solution.

Can we get brew to install this cask in ANDROID_HOME and then fail if ANDROID_HOME is not defined?

**Comment by tir38** [comment]
Looks like I can't change [where brew installs this one package](https://stackoverflow.com/a/31635024/1650674).

It also seems like setting `ANDROID_HOME` and `ANDROID_SDK_ROOT` to the Caskroom doesn't help:

```
$ echo $ANDROID_HOME
/opt/homebrew/Caskroom/android-commandlinetools/7302050
$ echo $ANDROID_SDK_ROOT/
/opt/homebrew/Caskroom/android-commandlinetools/7302050/
$ sdkmanager --list
Error: Could not determine SDK root.
Error: Either specify it explicitly with --sdk_root= or move this package into its expected location: <sdk>/cmdline-tools/latest/
```

The only thing I can think of is to
1. use brew to install command line tools into `/opt/homebrew/Caskroom`
2. use sdk manager to try to (re)install commandline tools (hopefully in to $ANDROID_HOME
```
$ /opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/bin/sdkmanager \
  "cmdline-tools;latest" \
  --sdk_root="/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/"
```

But no such luck. I can't tell `sdkmanager` *where* to install things, so I just end up with a nested copy of commandline tools:

```
/opt/homebrew/Caskroom/android-commandlinetools/7302050/cmdline-tools/cmdline-tools/latest
```

I'm at a loss here. This can't be the first time an app has required a specific install location.

**Comment by cho-m** [comment]
There doesn't seem to be an easy solution to this.
It will require a bit of `preflight/postflight` processing to work around the existing logic.

Some issues:
- Using the installation Caskroom as SDK root is probably a bad idea.
   When updates happen, it can get wiped out, which means losing anything installed via the command-line tools.
- A problem with the command scripts is that they have some logic to extract the (non-symlink) realpath.
   This means that using the easiest `postflight` symlink logic like in `android-sdk` Cask won't work.
- The commands don't seem to work with any environment variables.

---

Some options off the top of my head (there may be some better alternatives outside of these):

1. Change Caskroom installation structure to match expected path, like 
    ```
    $HOMEBREW_PREFIX/Caskroom/android-commandlinetools/<VERSION>/cmdline-tools/latest
    ```
    This will make the commands think `$HOMEBREW_PREFIX/Caskroom/android-commandlinetools/<VERSION>` is the SDK root. Then set the `--sdk_root` to some other directory, either as manual requirement or by creating a bunch of shim scripts (in order to avoid the 1st problem).

2. Copy the installed files into a separate directory (the 2nd problem requires these be realpaths and not symlinks) like:
    ```
    $HOMEBREW_PREFIX/share/android-sdk/cmdline-tools/latest
    ```
    This should probably make `$HOMEBREW_PREFIX/share/android-sdk` the root.
    The demerits here would be extra copy/delete logic for upgrades.

**Comment by cho-m** [comment]
@tir38 Did some basic updates to get something working. You can take a look, modify if necessary, and try it out.

@vitorgalvao If you have the chance, please review this Cask as successor to #80305 and see if there are any improvements to align to Homebrew-Cask requirements.

I don't use this Android CLT, so I cannot comment much on expected behavior.
The general points for what I modified:
- Using Caskroom as SDK root is probably bad due to wiping out anything installed on update.
   General idea I used is to copy into `share` direction (similar to #101021) as default.
- The commands expect to exist inside a directory like `.../cmdline-tools/latest/bin/<command>`
- The previous deprecated `android-sdk` had done symlinking, but this cask needs to actually copy over files due to realpath extraction (i.e. relating to previous point on directory structure). 
- Not very familiar with usage of this cask.
   Just included a `caveat` with exact location of installation in-case it is necessary.

The only things I confirmed were basic install/uninstall/zap and a simple commands like
```console
❯ sdkmanager --list
[=======================================] 100% Computing updates...
Available Packages:
  Path                                                                                     | Version    | Description
  -------                                                                                  | -------    | -------
  add-ons;addon-google_apis-google-15                                                      | 3          | Google APIs
...

❯ sdkmanager "platform-tools"
License android-sdk-license:            ] 10% Computing updates...
---------------------------------------
Terms and Conditions

This is the Android Software Development Kit License Agreement

1. Introduction
...

January 16, 2019
---------------------------------------
Accept? (y/N): y
[=======================================] 100% Unzipping... platform-tools/systr


❯ sdkmanager --list_installed
Installed packages:=====================] 100% Fetch remote repository...
  Path           | Version | Description                | Location
  -------        | ------- | -------                    | -------
  platform-tools | 31.0.2  | Android SDK Platform-Tools | platform-tools/

```

File: Casks/android-commandlinetools.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+cask "android-commandlinetools" do
+  version "7302050"
+  sha256 "fda8189832e506a58643c119f02c515a5a85741ae9c040fab41ee6c5ac021311"
+
+  url "https://dl.google.com/android/repository/commandlinetools-mac-#{version}_latest.zip",
+      verified: "dl.google.com/android/repository/"
+  name "android-cmdline-tools"
```
Comment (vitorgalvao) [inline Casks/android-commandlinetools.rb:7]: This is supposed to be a proper name. Like `Android Command-line Tools`.

File: Casks/android-commandlinetools.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    regex(%r{href=.*?/commandlinetools[._-]mac[._-](\d+)[._-]latest\.zip}i)
+  end
+
+  conflicts_with cask:    "android-sdk",
+                 formula: "proguard"
+
+  android_sdk_root = HOMEBREW_PREFIX.join("share", "android-commandlinetools")
+  android_clt_dir = android_sdk_root.join("cmdline-tools", "latest")
+
+  commands = %w[apkanalyzer avdmanager lint retrace screenshot2 sdkmanager]
```
Comment (vitorgalvao) [inline Casks/android-commandlinetools.rb:22]: These aren’t that many. Please just use `binary` multiple times.

File: Casks/android-commandlinetools.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  commands.each do |command|
+    binary "#{android_clt_dir}/bin/#{command}"
+  end
+
+  artifact "cmdline-tools", target: android_clt_dir
+
+  zap trash: android_sdk_root
```
Comment (vitorgalvao) [inline Casks/android-commandlinetools.rb:29]: Why is this a `zap` and not an `uninstall`?

**Review by vitorgalvao — CHANGES_REQUESTED** [review]

**Comment by vitorgalvao** [comment]
I will admit I’m getting tired of reviewing this cask. Why the insistence in getting this in? We keep having to bend over backwards to make this work and it doesn’t feel worth it at all. Hadn’t @cho-m spent time fixing this, I’d be more inclined to refuse it again.

I’m also not an Android developer so I can’t attest to how/if this will work.

File: Casks/android-commandlinetools.rb (hunk 6 above)
Comment (cho-m) [reply Casks/android-commandlinetools.rb:30]: The potential problem I saw was that Android CLT installs SDKs/etc into this directory. Running a command like upgrade/reinstall ends up wiping out those, which I am not sure is wanted.

Also causes issues due to order of uninstall:
```
==> Trashing files:
/opt/homebrew/share/android-commandlinetools
==> Purging files for version 7302050 of Cask android-commandlinetools
Error: android-commandlinetools: It seems the Generic Artifact source '/opt/homebrew/share/android-commandlinetools/cmdline-tools/latest' is not there.
```

**Comment by cho-m** [comment]
I also don't develop on Android, so I can't comment about usage/behavior.

I was thinking to at least give the Cask a trial period since it has at least some people interested.

Or someone can take the current work, modify it, and host it on their own tap.

From a cask standpoint, I can confirm:
- install/upgrade/reinstall/uninstall/zap works
- basic commands like `sdkmanager "platform-tools"` work out of the box

Still need someone who will actually use this Cask to confirm all basic behavior is good enough.

**Comment by cadenkriese** [comment]
> Still need someone who will actually use this Cask to confirm all basic behavior is good enough.

I just tried this out on my setup, performs exactly the same as the SDK when installed from android studio, I was able to get the emulator up and running with a sample app. 

One slight issue is when you install the `commandline-tools` package from `sdkmanager`, it gets confused having it installed twice as one might expect, could be worth adding a caveat message for. Other than that though everything seems to work as expected.

**Comment by tir38** [comment]
Update: I guess people can install `adb` directly from [platform-tools](https://formulae.brew.sh/cask/android-platform-tools#default) which is *not* deprecated

still it would be nice to have these tools available w/out having to install Android Studio

**Comment by cho-m** [comment]
> One slight issue is when you install the `commandline-tools` package from `sdkmanager`

How often does a user do this?

It is possible to install artifact into a fake directory like
`#{HOMEBREW_PREFIX}/share/android-commandlinetools/cmdline-tools/homebrew`
It currently seems to work as the tools don't seem to care if directory is called `latest`.
Though, it may break in a future release of Android CLT.
This would avoid the common `latest`, `4.0`, `3.0`, etc. names

**Comment by vitorgalvao** [comment]
@cho-m I’m fine with this whenever you are.

**Comment by cho-m** [comment]
@vitorgalvao Okay. Just want to get user-input for final commits.

@cadenkriese, @tir38, and anyone else interested, please provide any final input whether latest changes to Cask are working okay. I ended up adding a fake version "`homebrew`" to avoid issue previously mentioned when `sdkmanager` installs a copy of `cmdline-tools;latest`

**Comment by cadenkriese** [comment]
> @vitorgalvao Okay. Just want to get user-input for final commits.
> 
> @cadenkriese, @tir38, and anyone else interested, please provide any final input whether latest changes to Cask are working okay. I ended up adding a fake version "`homebrew`" to avoid issue previously mentioned when `sdkmanager` installs a copy of `cmdline-tools;latest`

That seems like a great workaround, I don't see any issues with my workflow. Thanks for sticking with this cask!

**Comment by cho-m** [comment]
@cadenkriese Thanks for testing Cask out

@vitorgalvao Might as well merge this now.
If users see any issues, I can help look into them.

**Review by miccal — APPROVED** [review]

**Comment by tir38** [comment]
Thanks y'all. I'm setting up a new machine this weekend. So I'll have a chance to test this out on a fresh machine.

**Comment by igoriuz** [comment]
First: Thanks your work! 

How do you link it this new "android sdk" to Android Studio? As far as i can tell, android studio intially doesn't care about environment variables as it's looking for `~/Library/Android/sdk`. My current workaround is by creating Android folder and a symlink to android-commandlinetools path. Do you guys have better ideas? Or can we integrate this somehow into the formulae?

**Pull request merged**